### PR TITLE
virsh.migrate: Add one case for timeout-suspend

### PR DESCRIPTION
**Pull request by dzhengfy**
Signed-off-by: Dan Zheng <dzheng@redhat.com>

### CODE REVIEW

**Comment by dzhengfy** [comment]
This depends on https://github.com/avocado-framework/avocado-vt/pull/648 and https://github.com/avocado-framework/avocado-vt/pull/645

**Comment by dzhengfy** [comment]
# avocado run --vt-type libvirt virsh.migrate.there_timeout_suspend
JOB ID     : 7f3582107d97eb6566aa347b49099406d24cebf8
JOB LOG    : /root/avocado/job-results/job-2016-08-05T17.26-7f35821/job.log
TESTS      : 1
 (1/1) type_specific.io-github-autotest-libvirt.virsh.migrate.there_timeout_suspend: PASS (215.43 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0
JOB HTML   : /root/avocado/job-results/job-2016-08-05T17.26-7f35821/html/results.html
TESTS TIME : 215.43 s

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 1]
```diff
@@ -634,7 +637,62 @@ def hugepage_assign(hp_num, target_ip='', node='', hp_size='', user='',
...
+        # 3. Start the migration
+        # 4. When the eclipse time reaches the timeout value, check the guest
+        #    state to be paused on both source host and target host
+        # 5. Wait for the migration done. Check the guest state to be shutoff
+        #    on source host and running on target host
+        if extra.count("--timeout-suspend"):
+            asynch_migration = True
+            speed = int(params.get("migrate_speed", 1))
+            timeout = int(params.get("timeout_before_suspend", 5))
+            logging.debug("To set the migration speed '%s'M", speed)
```
Comment (will-Do) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: "Setting migration speed to %sM"

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 2]
```diff
@@ -634,7 +637,62 @@ def hugepage_assign(hp_num, target_ip='', node='', hp_size='', user='',
...
+            virsh.migrate_setspeed(vm_name, speed, debug=True)
+            cp_mig = libvirt.MigrationTest()
```
Comment (will-Do) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Maybe you can find a better name for this

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 3]
```diff
@@ -634,7 +637,62 @@ def hugepage_assign(hp_num, target_ip='', node='', hp_size='', user='',
...
+            migrate_options = "%s %s" % (options, extra)
+            mig_thread, stime = cp_mig.do_migration_asynch(vm, None, dest_uri,
+                                                           migrate_options,
+                                                           ignore_status=True)
+            ssh_auto_recover = False
+            logging.debug("To wait for %s seconds", timeout)
```
Comment (will-Do) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Okay, I just find all these debug messages in this patch start with 'To ...', maybe you could rewrite them in a better way.

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 4]
```diff
@@ -634,7 +637,62 @@ def hugepage_assign(hp_num, target_ip='', node='', hp_size='', user='',
...
+            # Add buffer seconds to avoid of timing gap
+            time.sleep(timeout + 3)
```
Comment (will-Do) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: So the buffer here is 3 sec? could you please add more explanation above

**Comment by dzhengfy** [comment]
@will-Do, updated for your comments. Please double check. Thanks.
@kylazhang , could you help review it, too?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 5]
```diff
@@ -634,14 +682,47 @@ def hugepage_assign(hp_num, target_ip='', node='', hp_size='', user='',
...
+        if extra.count("--timeout-suspend"):
+            asynch_migration = True
+            speed = int(params.get("migrate_speed", 1))
+            timeout = int(params.get("timeout_before_suspend", 5))
+            logging.debug("Set migration speed to %sM", speed)
+            virsh.migrate_setspeed(vm_name, speed, debug=True)
+            migration_test = libvirt.MigrationTest()
+            migrate_options = "%s %s" % (options, extra)
+            vms = [vm]
+            ssh_auto_recover = False
```
Comment (kylazhang) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Why do not need auto_recover? Could you add comments here?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py [hunk 6]
```diff
@@ -696,6 +777,16 @@ def hugepage_assign(hp_num, target_ip='', node='', hp_size='', user='',
...
 
+        # Check vm state on source.
+        if extra.count("--timeout-suspend"):
+            logging.debug("Checking '%s' state on source '%s'", vm.name,
+                          src_uri)
+            vm_state = virsh.domstate(vm.name, uri=src_uri).stdout.strip()
+            if vm_state != "shut off":
+                raise exceptions.TestFail("Local vm state should be '%s', "
+                                          "but found '%s'"
+                                          % ('shut off', vm_state))
```
Comment (kylazhang) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: Why do not direct write 'shut off'?

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 5 above)
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: remove it.

File: libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py (hunk 6 above)
Comment (dzhengfy) [inline libvirt/tests/src/virsh_cmd/domain/virsh_migrate.py]: update it.

**Comment by dzhengfy** [comment]
@kylazhang , updated for your comments. Please help double check thanks.

**Comment by kylazhang** [comment]
ACK. @waynesun09 and @Hao-Liu please help to merge. Thanks

**Comment by dzhengfy** [comment]
@kylazhang @waynesun09 , could you help review?

**Comment by waynesun09** [comment]
ACK and merge now, thanks

**Pull request merged**